File: app/src/main/java/dev/ragnarok/filegallery/api/impl/LocalServerApi.java
package dev.ragnarok.filegallery.api.impl;

import static dev.ragnarok.filegallery.util.Objects.nonNull;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.filegallery.api.ILocalServerServiceProvider;
import dev.ragnarok.filegallery.api.interfaces.ILocalServerApi;
import dev.ragnarok.filegallery.api.model.Items;
import dev.ragnarok.filegallery.api.model.response.BaseResponse;
import dev.ragnarok.filegallery.model.Audio;
import dev.ragnarok.filegallery.model.Photo;
import dev.ragnarok.filegallery.model.Video;
import dev.ragnarok.filegallery.util.Utils;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.exceptions.Exceptions;
import io.reactivex.rxjava3.functions.Function;

class LocalServerApi implements ILocalServerApi {

    private final ILocalServerServiceProvider service;

    LocalServerApi(ILocalServerServiceProvider service) {
        this.service = service;
    }

    static <T> Function<BaseResponse<Items<T>>, List<T>> extractResponseWithErrorHandling() {
        return response -> {
            if (nonNull(response.error)) {
                throw Exceptions.propagate(new Exception(Utils.firstNonEmptyString(response.error.errorMsg, "Error")));
            }
            if (response.response == null || Utils.isEmpty(response.response.items)) {
                return new ArrayList<>();
            }
            return response.response.items;
        };
    }

    static <T> Function<BaseResponse<T>, T> extractResponseWithErrorHandlingSimple() {
        return response -> {
            if (nonNull(response.error)) {
                throw Exceptions.propagate(new Exception(Utils.firstNonEmptyString(response.error.errorMsg, "Error")));
            }

            return response.response;
        };
    }

    @Override
    public Single<List<Video>> getVideos(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getVideos(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Audio>> getAudios(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getAudios(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Photo>> getPhotos(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getPhotos(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Audio>> getDiscography(Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.getDiscography(offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Video>> searchVideos(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchVideos(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Photo>> searchPhotos(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchPhotos(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Audio>> searchAudios(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchAudios(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<List<Audio>> searchDiscography(String query, Integer offset, Integer count, boolean reverse) {
        return service.provideLocalServerService()
                .flatMap(service -> service.searchDiscography(query, offset, count, reverse ? 1 : 0)
                        .map(extractResponseWithErrorHandling()));
    }

    @Override
    public Single<Integer> update_time(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.update_time(hash)
                        .map(extractResponseWithErrorHandlingSimple()));
    }

    @Override
    public Single<Integer> delete_media(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.delete_media(hash)
                        .map(extractResponseWithErrorHandlingSimple()));
    }

    @Override
    public Single<String> get_file_name(String hash) {
        return service.provideLocalServerService()
                .flatMap(service -> service.get_file_name(hash)
                        .map(extractResponseWithErrorHandlingSimple()));
    }

    @Override
    public Single<Integer> update_file_name(String hash, String name) {
        return service.provideLocalServerService()
                .flatMap(service -> service.update_file_name(hash, name)
                        .map(extractResponseWithErrorHandlingSimple()));
    }
}

File: app/src/main/java/dev/ragnarok/filegallery/api/impl/LocalServerApi.kt
package dev.ragnarok.filegallery.api.impl

import dev.ragnarok.filegallery.api.ILocalServerServiceProvider
import dev.ragnarok.filegallery.api.interfaces.ILocalServerApi
import dev.ragnarok.filegallery.api.model.Items
import dev.ragnarok.filegallery.api.model.response.BaseResponse
import dev.ragnarok.filegallery.api.services.ILocalServerService
import dev.ragnarok.filegallery.model.Audio
import dev.ragnarok.filegallery.model.Photo
import dev.ragnarok.filegallery.model.Video
import dev.ragnarok.filegallery.nonNullNoEmpty
import dev.ragnarok.filegallery.util.Utils.firstNonEmptyString
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.exceptions.Exceptions
import io.reactivex.rxjava3.functions.Function

internal class LocalServerApi(private val service: ILocalServerServiceProvider) : ILocalServerApi {
    override fun getVideos(offset: Int?, count: Int?, reverse: Boolean): Single<List<Video>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.getVideos(offset, count, if (reverse) 1 else 0)
                    .map<List<Video>>(extractResponseWithErrorHandling())
            }
    }

    override fun getAudios(offset: Int?, count: Int?, reverse: Boolean): Single<List<Audio>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.getAudios(offset, count, if (reverse) 1 else 0)
                    .map<List<Audio>>(extractResponseWithErrorHandling())
            }
    }

    override fun getPhotos(
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Single<MutableList<Photo>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.getPhotos(offset, count, if (reverse) 1 else 0)
                    .map(extractResponseWithErrorHandlingMutable())
            }
    }

    override fun getDiscography(offset: Int?, count: Int?, reverse: Boolean): Single<List<Audio>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.getDiscography(offset, count, if (reverse) 1 else 0)
                    .map<List<Audio>>(extractResponseWithErrorHandling())
            }
    }

    override fun searchVideos(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Single<List<Video>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.searchVideos(query, offset, count, if (reverse) 1 else 0)
                    .map<List<Video>>(extractResponseWithErrorHandling())
            }
    }

    override fun searchPhotos(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Single<List<Photo>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.searchPhotos(query, offset, count, if (reverse) 1 else 0)
                    .map<List<Photo>>(extractResponseWithErrorHandling())
            }
    }

    override fun searchAudios(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Single<List<Audio>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.searchAudios(query, offset, count, if (reverse) 1 else 0)
                    .map<List<Audio>>(extractResponseWithErrorHandling())
            }
    }

    override fun searchDiscography(
        query: String?,
        offset: Int?,
        count: Int?,
        reverse: Boolean
    ): Single<List<Audio>> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.searchDiscography(query, offset, count, if (reverse) 1 else 0)
                    .map<List<Audio>>(extractResponseWithErrorHandling())
            }
    }

    override fun update_time(hash: String?): Single<Int> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.update_time(hash)
                    .map(extractResponseWithErrorHandlingSimple())
            }
    }

    override fun delete_media(hash: String?): Single<Int> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.delete_media(hash)
                    .map(extractResponseWithErrorHandlingSimple())
            }
    }

    override fun get_file_name(hash: String?): Single<String> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.get_file_name(hash)
                    .map(extractResponseWithErrorHandlingSimple())
            }
    }

    override fun update_file_name(hash: String?, name: String?): Single<Int> {
        return service.provideLocalServerService()
            .flatMap { service: ILocalServerService ->
                service.update_file_name(hash, name)
                    .map(extractResponseWithErrorHandlingSimple())
            }
    }

    companion object {
        inline fun <reified T> extractResponseWithErrorHandling(): Function<BaseResponse<Items<T>?>, List<T>> {
            return Function { response: BaseResponse<Items<T>?> ->
                if (response.error != null) {
                    throw Exceptions.propagate(
                        Exception(
                            firstNonEmptyString(
                                response.error.errorMsg,
                                "Error"
                            )
                        )
                    )
                }
                response.response?.items.nonNullNoEmpty {
                    return@Function it
                }
                return@Function ArrayList<T>()
            }
        }

        inline fun <reified T> extractResponseWithErrorHandlingMutable(): Function<BaseResponse<Items<T>?>, MutableList<T>> {
            return Function { response: BaseResponse<Items<T>?> ->
                if (response.error != null) {
                    throw Exceptions.propagate(
                        Exception(
                            firstNonEmptyString(
                                response.error.errorMsg,
                                "Error"
                            )
                        )
                    )
                }
                response.response?.items.nonNullNoEmpty {
                    return@Function it
                }
                return@Function ArrayList<T>()
            }
        }

        inline fun <reified T : Any> extractResponseWithErrorHandlingSimple(): Function<BaseResponse<T>, T> {
            return Function { response: BaseResponse<T> ->
                if (response.error != null) {
                    throw Exceptions.propagate(
                        Exception(
                            firstNonEmptyString(
                                response.error.errorMsg,
                                "Error"
                            )
                        )
                    )
                }
                response.response
            }
        }
    }
}